### ISSUE TITLE
Assertion `!R.empty() && (*R.begin())->isCXXClassMember()' failed on invalid C++ code

### ISSUE THREAD

**Issue by mstorsjo** [llvm/llvm-project#40427]
|  |  |
| --- | --- |
| Bugzilla Link | [41082](https://llvm.org/bz41082) |
| Version | trunk |
| OS | All |
| CC | @aaronpuchert,@DougGregor,@zygoloid |

## Extended Description 
If trying to compile the following snippet of broken C++ code, clang triggers a failed assert:

$ cat clang-sema-assert.cpp 

```cpp
struct A {
  memmove;
  namespace using ::memmove;
    a ( memmove
```

$ clang++ -c clang-sema-assert.cpp 
clang-sema-assert.cpp:2:3: error: C++ requires a type specifier for all
      declarations
  memmove;
  ^
clang-sema-assert.cpp:1:8: error: missing '}' at end of definition of 'A'
struct A {
       ^
clang-sema-assert.cpp:3:3: note: still within definition of 'A' here
  namespace using ::memmove;
  ^
clang-sema-assert.cpp:3:13: error: expected identifier or '{'
  namespace using ::memmove;
            ^
clang-sema-assert.cpp:3:19: error: no member named 'memmove' in the global
      namespace; did you mean 'A::memmove'?
  namespace using ::memmove;
                  ^~~~~~~~~
                  A::memmove
clang-sema-assert.cpp:2:3: note: 'A::memmove' declared here
  memmove;
  ^
clang-9: ../tools/clang/lib/Sema/SemaExprMember.cpp:90: IMAKind ClassifyImplicitMemberAccess(clang::Sema&, const clang::LookupResult&): Assertion `!R.empty() && (*R.begin())->isCXXClassMember()' failed.
Stack dump:
0.	Program arguments: /home/martin/code/llvm/build/bin/clang-9 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all -disable-free -main-file-name clang-sema-assert.cpp -mrelocation-model static -mthread-model posix -mdisable-fp-elim -fmath-errno -masm-verbose -mconstructor-aliases -munwind-tables -fuse-init-array -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -coverage-notes-file /home/martin/code/llvm/build/clang-sema-assert.gcno -resource-dir /home/martin/code/llvm/build/lib/clang/9.0.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/backward -internal-isystem /usr/local/include -internal-isystem /home/martin/code/llvm/build/lib/clang/9.0.0/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir /home/martin/code/llvm/build -ferror-limit 19 -fmessage-length 80 -fobjc-runtime=gcc -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -o clang-sema-assert.o -x c++ clang-sema-assert.cpp -faddrsig 
1.	clang-sema-assert.cpp:4:9: current parser token 'memmove'
 #&#8203;0 0x0000000001ae5a2a llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/home/martin/code/llvm/build/bin/clang-9+0x1ae5a2a)
 #&#8203;1 0x0000000001ae37cc llvm::sys::RunSignalHandlers() (/home/martin/code/llvm/build/bin/clang-9+0x1ae37cc)
 #&#8203;2 0x0000000001ae3937 SignalHandler(int) (/home/martin/code/llvm/build/bin/clang-9+0x1ae3937)
 #&#8203;3 0x00007f689e927390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
 #&#8203;4 0x00007f689d699428 gsignal /build/glibc-LK5gWL/glibc-2.23/signal/../sysdeps/unix/sysv/linux/raise.c:54:0
 #&#8203;5 0x00007f689d69b02a abort /build/glibc-LK5gWL/glibc-2.23/stdlib/abort.c:91:0
 #&#8203;6 0x00007f689d691bd7 __assert_fail_base /build/glibc-LK5gWL/glibc-2.23/assert/assert.c:92:0
 #&#8203;7 0x00007f689d691c82 (/lib/x86_64-linux-gnu/libc.so.6+0x2dc82)
 #&#8203;8 0x0000000002d4c3ea clang::Sema::BuildPossibleImplicitMemberExpr(clang::CXXScopeSpec const&, clang::SourceLocation, clang::LookupResult&, clang::TemplateArgumentListInfo const*, clang::Scope const*) (/home/martin/code/llvm/build/bin/clang-9+0x2d4c3ea)
 #&#8203;9 0x0000000002bb2320 clang::Sema::ClassifyName(clang::Scope*, clang::CXXScopeSpec&, clang::IdentifierInfo*&, clang::SourceLocation, clang::Token const&, bool, std::unique_ptr<clang::CorrectionCandidateCallback, std::default_delete<clang::CorrectionCandidateCallback> >) (/home/martin/code/llvm/build/bin/clang-9+0x2bb2320)
#&#8203;10 0x0000000002999fd6 clang::Parser::TryAnnotateName(bool, std::unique_ptr<clang::CorrectionCandidateCallback, std::default_delete<clang::CorrectionCandidateCallback> >) (/home/martin/code/llvm/build/bin/clang-9+0x2999fd6)
#&#8203;11 0x0000000002a2fa7d clang::Parser::isCXXDeclarationSpecifier(clang::Parser::TPResult, bool*) (/home/martin/code/llvm/build/bin/clang-9+0x2a2fa7d)
#&#8203;12 0x0000000002a33228 clang::Parser::TryParseParameterDeclarationClause(bool*, bool) (/home/martin/code/llvm/build/bin/clang-9+0x2a33228)
#&#8203;13 0x0000000002a336bd clang::Parser::isCXXFunctionDeclarator(bool*) (/home/martin/code/llvm/build/bin/clang-9+0x2a336bd)
#&#8203;14 0x00000000029bd13d clang::Parser::ParseDirectDeclarator(clang::Declarator&) (/home/martin/code/llvm/build/bin/clang-9+0x29bd13d)
#&#8203;15 0x00000000029ad967 clang::Parser::ParseDeclaratorInternal(clang::Declarator&, void (clang::Parser::*)(clang::Declarator&)) (/home/martin/code/llvm/build/bin/clang-9+0x29ad967)
#&#8203;16 0x00000000029ade12 clang::Parser::ParseDeclarator(clang::Declarator&) (/home/martin/code/llvm/build/bin/clang-9+0x29ade12)
#&#8203;17 0x00000000029b7672 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/home/martin/code/llvm/build/bin/clang-9+0x29b7672)
#&#8203;18 0x000000000299349c clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/home/martin/code/llvm/build/bin/clang-9+0x299349c)
#&#8203;19 0x0000000002993cb1 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (.part.220.constprop.223) (/home/martin/code/llvm/build/bin/clang-9+0x2993cb1)
#&#8203;20 0x0000000002993d0f clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/home/martin/code/llvm/build/bin/clang-9+0x2993d0f)
#&#8203;21 0x0000000002998620 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/home/martin/code/llvm/build/bin/clang-9+0x2998620)
#&#8203;22 0x0000000002998f1f clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) (/home/martin/code/llvm/build/bin/clang-9+0x2998f1f)
#&#8203;23 0x000000000298f6b3 clang::ParseAST(clang::Sema&, bool, bool) (/home/martin/code/llvm/build/bin/clang-9+0x298f6b3)
#&#8203;24 0x0000000002620faf clang::CodeGenAction::ExecuteAction() (/home/martin/code/llvm/build/bin/clang-9+0x2620faf)
#&#8203;25 0x000000000214cb3e clang::FrontendAction::Execute() (/home/martin/code/llvm/build/bin/clang-9+0x214cb3e)
#&#8203;26 0x0000000002114c06 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/martin/code/llvm/build/bin/clang-9+0x2114c06)
#&#8203;27 0x00000000021eb03c clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/martin/code/llvm/build/bin/clang-9+0x21eb03c)
#&#8203;28 0x0000000000940c88 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/martin/code/llvm/build/bin/clang-9+0x940c88)
#&#8203;29 0x00000000008d1219 main (/home/martin/code/llvm/build/bin/clang-9+0x8d1219)
#&#8203;30 0x00007f689d684830 __libc_start_main /build/glibc-LK5gWL/glibc-2.23/csu/../csu/libc-start.c:325:0
#&#8203;31 0x000000000093cb59 _start (/home/martin/code/llvm/build/bin/clang-9+0x93cb59)

**Comment by shafik** [comment]
It looks like the original case does not crash anymore: https://godbolt.org/z/jsxhdz65x

**Comment by mstorsjo** [comment]
> It looks like the original case does not crash anymore: https://godbolt.org/z/jsxhdz65x
> 
> nor does the last example either.

Thanks! A quick bisect shows that it seems like the issue was fixed by 23d6525cbdc9de7cbfe7640d1e9e4f25a0c5dd85:
```
commit 23d6525cbdc9de7cbfe7640d1e9e4f25a0c5dd85
Author: Richard Smith <richard@metafoo.co.uk>
Date:   Mon Jul 27 17:28:06 2020 -0700

    Don't form a 'context-independent expr' reference to a member during
    name annotation.
    
    Instead, defer forming the member access expression or DeclRefExpr until
    we build the use of ClassifyName's result. Just build an
    UnresolvedLookupExpr to track the LookupResult until we're ready to
    consume it.
    
    This also reverts commit 2f7269b6773de2750f9cd1417ef5f21cd6cf7a91 (other
    than its testcase). That change was an attempted workaround for the same
    problem.

```
